### PR TITLE
finished day13

### PR DESCRIPTION
**Pull request by KoenvdBerg**
Hi JP,

For this puzzle I took a lot of inspiration from the solution of nmcb, because it was part of the curriculum. 

It works by first parsing in the input file with a Parser Combinator. While working on chapter09 of the FPS book, I wrote my own parser combinator library. This is the library that I've included and used here. 

To sort the packets, I defined an ordering for sorting the parsed packets. The remainder of the puzzle then was trivial. 

Curious what you think of the parser combinator library that I've created.

### CODE REVIEW

**Review by jan-pieter — COMMENTED** [review]
The parser lib looks quite powerful. The hard thing with parser libraries is producing nice error messages which I think is hard to implement :wink:

File: aoc2022/src/main/scala/day13.scala [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    val packet: Parser[Packet] = n | l
+    def parsePacket(s: String): Packet = packet.run(s) match
+      case Right(v) => v
+      case Left(e) => sys.error(e)
+    
+    given ordering: Ordering[Packet] with
+      override def compare(x: Packet, y: Packet): Int = (x, y) match
+        case (N(i1), N(i2))       => i1.compare(i2)
+        case (n@N(_), l@L(_))     => compare(L(List(n)), l)
+        case (l@L(_), n@N(_))     => compare(l, L(List(n)))
```
[the remark this answers is not in the document]
Comment (nmcb) [reply aoc2022/src/main/scala/day13.scala:63]: true - this particular lib has a `run[A] : String Option[(A,String)]` type - which basically means that it provides _no_ facility to implement error messages anyhow - we could go towards a "simple" `run[A] : String => Either[String,(A,String)]` type to start providing rudimentary error messages tho <3

[the remark this answers is not in the document]
Comment (nmcb) [reply aoc2022/src/main/scala/day13.scala:63]: true - this particular lib has a `run[A] : String => Option[(A,String)]` type - which basically means that it provides _no_ facility to implement error messages anyhow - we could go towards a "simple" `run[A] : String => Either[String,(A,String)]` type to start providing rudimentary error messages tho <3

File: aoc2022/src/main/scala/day13.scala [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  println(s"Answer day $day part 1: ${answer1} [${System.currentTimeMillis - start1}ms]")
+
+  private val start2: Long =
+    System.currentTimeMillis
+
+  private val div1: Packet = Packet.parsePacket("[[2]]")  
+  private val div2: Packet = Packet.parsePacket("[[6]]")  
+  
+  private val newInput: List[Packet] = (div1 :: div2 :: input).sorted
+  private val answer2: Int = (newInput.indexWhere(_ == div1) + 1) * (newInput.indexWhere(_ == div2) + 1)
```
[the remark this answers is not in the document]
Comment (KoenvdBerg) [reply aoc2022/src/main/scala/day13.scala:87]: Yes it worked. Didn't know `indexOf` existed the time I wrote this code. Thanks!

**Pull request merged**